### PR TITLE
Generalize toAccumulateType()

### PR DESCRIPTION
**Pull request by qqaatw**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108248

Trying to address this comment: https://github.com/pytorch/pytorch/pull/106666#discussion_r1297397554

cc @albanD @kulinseth

### CODE REVIEW

File: aten/src/ATen/AccumulateType.h [hunk 1]
```diff
@@ -51,147 +52,231 @@
...
+  using type = void;
+};
+template <>
+struct AccumulateTypeDevice<c10::complex<float>, c10::DeviceType::MPS> {
+  using type = void;
+};
+template <>
+struct AccumulateTypeDevice<c10::complex<double>, c10::DeviceType::MPS> {
+  using type = void;
+};
```
Comment (qqaatw) [inline aten/src/ATen/AccumulateType.h:121]: Not sure what's the best way to assert these currently unsupported dtypes.

File: aten/src/ATen/AccumulateType.h [hunk 2]
```diff
@@ -51,147 +52,231 @@
 
 namespace at {
 
-template <typename T, bool is_cuda>
-struct AccumulateType {};
+template <typename T, c10::DeviceType D>
+struct AccumulateTypeDevice {};
+
+template <>
```
Comment (albanD) [inline aten/src/ATen/AccumulateType.h:58]: nit: I would use macro to make these definitions more compact and easier to read with something like:
```
MPS_ACC_TYPE(BFloat16, float);
MPS_ACC_TYPE(Half, float);
...
MPS_ACC_TYPE(int8_t, int64_t);
...
```

File: aten/src/ATen/AccumulateType.h (hunk 1 above)
Comment (albanD) [reply aten/src/ATen/AccumulateType.h:121]: Good question!
What happens in the void case here?
Otherwise not defining type should give a compilation error if you try to use this version of the template?

**Review by albanD — COMMENTED** [review]
Super cool update!

File: aten/src/ATen/AccumulateType.h (hunk 2 above)
Comment (qqaatw) [reply aten/src/ATen/AccumulateType.h]: Love this suggestion! Done.

File: aten/src/ATen/AccumulateType.h (hunk 1 above)
Comment (qqaatw) [reply aten/src/ATen/AccumulateType.h]: I was experimenting on whether assigning `void` here could give us some kind of signal (like a dummy value) indicating this type is not supported, but it causes a compilation error indeed. So this is not viable.

Do you have suggestions on this?

Comment (albanD) [reply aten/src/ATen/AccumulateType.h]: Isn't compilation error ok? People should never use that!

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 3]
```diff
@@ -1624,11 +1623,11 @@ Tensor renorm_jvp(
...
-      factor *
-          (self_t -
-           self_p * invnorm *
-               norm_jvp(
-                   self_p, self_t, p, norm, reduce_dims, /*keepdim=*/true)),
+      (factor *
+       (self_t -
+        self_p * invnorm *
+            norm_jvp(self_p, self_t, p, norm, reduce_dims, /*keepdim=*/true)))
+          .to(self_t.scalar_type()),
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:1630]: Why this change is needed now?

File: aten/src/ATen/AccumulateType.h [hunk 4]
```diff
@@ -51,147 +52,94 @@
...
+#define ACC_TYPE(t, acc_t, device_type)         \
+  template <>                                   \
+  struct AccumulateTypeDevice<t, device_type> { \
+    using type = acc_t;                         \
+  };
+#define MPS_ACC_TYPE(t, acc_t) ACC_TYPE(t, acc_t, c10::DeviceType::MPS)
+#define CUDA_ACC_TYPE(t, acc_t) ACC_TYPE(t, acc_t, c10::DeviceType::CUDA)
+#define CPU_ACC_TYPE(t, acc_t) ACC_TYPE(t, acc_t, c10::DeviceType::CPU)
+
+MPS_ACC_TYPE(BFloat16, float);
```
Comment (albanD) [inline aten/src/ATen/AccumulateType.h:86]: This looks great! Thanks for the refactor!

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 3 above)
Comment (qqaatw) [reply torch/csrc/autograd/FunctionsManual.cpp:1630]: The behaviors for renorm forward and backward are a bit different:

In forward, the `acc_type` is fixed to the ones for cuda:

https://github.com/pytorch/pytorch/blob/1aacbaed8b7b85afe93d039fd27e2b1d2f77bb8f/aten/src/ATen/native/Normalization.cpp#L858-L861

In backward, the `acc_type` depends on the device of the input tensor:

https://github.com/pytorch/pytorch/blob/1aacbaed8b7b85afe93d039fd27e2b1d2f77bb8f/torch/csrc/autograd/FunctionsManual.cpp#L1565-L1569

To be honest I'm not sure why the difference presents but I feel the backward one would be more appropriate. So I aligned the behavior of `acc_type` for `renorm_jvp` with the one in backward, and therefore if the `acc_type` is different from the original type, we should cast the type back, hence this change.

File: aten/src/ATen/AccumulateType.h (hunk 1 above)
Comment (qqaatw) [reply aten/src/ATen/AccumulateType.h]: Hmm, the compilation error causes the CI builds to fail: https://hud.pytorch.org/pytorch/pytorch/pull/108248?sha=87fcbd8ab0cbd89126dbe27bc873866ca6315ea5

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 3 above)
Comment (albanD) [reply torch/csrc/autograd/FunctionsManual.cpp:1630]: That sounds like a dangerous change we would need to review super carefully. I think it would be best to land the cleanup and do this change in a follow up. WDYT?

Comment (qqaatw) [reply torch/csrc/autograd/FunctionsManual.cpp]: Sure.

File: aten/src/ATen/AccumulateType.h (hunk 1 above)
Comment (qqaatw) [reply aten/src/ATen/AccumulateType.h]: The complex64 support was added recently, and complex128 is not yet supported I think.

https://github.com/pytorch/pytorch/blob/c3945b5f8496f068d56ea1879c10d62ee4ec1801/test/test_mps.py#L10816

Therefore, I set the complex acc types to complex64.

File: aten/src/ATen/AccumulateType.h [hunk 5]
```diff
@@ -51,147 +52,94 @@
...
+MPS_ACC_TYPE(int8_t, int64_t);
+MPS_ACC_TYPE(uint8_t, int64_t);
+MPS_ACC_TYPE(char, int64_t);
+MPS_ACC_TYPE(int16_t, int64_t);
+MPS_ACC_TYPE(int32_t, int64_t);
+MPS_ACC_TYPE(int64_t, int64_t);
+MPS_ACC_TYPE(bool, bool);
+MPS_ACC_TYPE(c10::complex<Half>, c10::complex<float>);
+MPS_ACC_TYPE(c10::complex<float>, c10::complex<float>);
+MPS_ACC_TYPE(c10::complex<double>, c10::complex<float>);
```
Comment (albanD) [inline aten/src/ATen/AccumulateType.h:101]: cc @kulinseth just to confirm that this is ok to have this for now.

**Review by albanD — COMMENTED** [review]
Change sounds good.
I just want a quick stamp from Kulin to make sure the double complex handling is fine.

File: aten/src/ATen/AccumulateType.h (hunk 5 above)
Comment (kulinseth) [reply aten/src/ATen/AccumulateType.h:101]: Can we issue a warning that we are falling back to using float for double types ?

**Review by kulinseth — APPROVED** [review]
Other than the warning . This looks good to me

File: aten/src/ATen/AccumulateType.h (hunk 5 above)
Comment (qqaatw) [reply aten/src/ATen/AccumulateType.h:101]: I'm a bit concerned about that would be somewhat verbose from a user's point of view since accumulate type is an internal thing of which a regular user probably doesn't have idea.

What do you think @albanD?

Comment (albanD) [reply aten/src/ATen/AccumulateType.h:101]: I'm not sure what is the best place to put this warning for sure. Also note that the user would need to have created a compleex<double> to begin with to reach this.
So I would say that not having a warning here is fine?

Comment (qqaatw) [reply aten/src/ATen/AccumulateType.h:101]: If the current state is fine, this PR needs your approval @albanD. Thanks!

**Review by albanD — APPROVED** [review]
SGTM
